# Install and import required libraries

In [1]:
import numpy as np
import pandas as pd
import json
!pip install geopy
from geopy.geocoders import Nominatim
import requests
from pandas.io.json import json_normalize
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
!pip install folium
import folium
!pip install beautifulsoup4
from bs4 import BeautifulSoup

print('Libraries imported.')

Libraries imported.


# Part 1: Scraping data from Wikipedia page

Create GET request URL.

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
url

'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

Send GET request.

In [3]:
results = requests.get(url)

Create BeautifulSoup object and examine its contents.

In [4]:
html = results.text
soup = BeautifulSoup(html)
print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"XrRLwQpAMMQAAeqP5OEAAABS","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":955414546,"wgRevisionId":955414546,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Communications in Ontario","Postal codes in Canada","Toronto","Ontario

The target table has __\<table>__ tag with class name "wikitable sortable".

In [5]:
table = soup.find('table', class_='wikitable sortable')
table

<table class="wikitable sortable">
<tbody><tr>
<th>Postal Code
</th>
<th>Borough
</th>
<th>Neighborhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park, Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor, Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park, Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern, Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3B
</td>
<td>North York
</td>
<td>Don Mills
</td></tr>
<tr>
<td>M4B
</td>
<td>East Y

In the table, tr tag defines a table row, th tag defines a table header, td tag defines a table cell. We will extract data from the table according to these tags.

In [6]:
# Find all rows with <tr> tag
rows = table.find_all('tr')
rows

[<tr>
 <th>Postal Code
 </th>
 <th>Borough
 </th>
 <th>Neighborhood
 </th></tr>, <tr>
 <td>M1A
 </td>
 <td>Not assigned
 </td>
 <td>
 </td></tr>, <tr>
 <td>M2A
 </td>
 <td>Not assigned
 </td>
 <td>
 </td></tr>, <tr>
 <td>M3A
 </td>
 <td>North York
 </td>
 <td>Parkwoods
 </td></tr>, <tr>
 <td>M4A
 </td>
 <td>North York
 </td>
 <td>Victoria Village
 </td></tr>, <tr>
 <td>M5A
 </td>
 <td>Downtown Toronto
 </td>
 <td>Regent Park, Harbourfront
 </td></tr>, <tr>
 <td>M6A
 </td>
 <td>North York
 </td>
 <td>Lawrence Manor, Lawrence Heights
 </td></tr>, <tr>
 <td>M7A
 </td>
 <td>Downtown Toronto
 </td>
 <td>Queen's Park, Ontario Provincial Government
 </td></tr>, <tr>
 <td>M8A
 </td>
 <td>Not assigned
 </td>
 <td>
 </td></tr>, <tr>
 <td>M9A
 </td>
 <td>Etobicoke
 </td>
 <td>Islington Avenue
 </td></tr>, <tr>
 <td>M1B
 </td>
 <td>Scarborough
 </td>
 <td>Malvern, Rouge
 </td></tr>, <tr>
 <td>M2B
 </td>
 <td>Not assigned
 </td>
 <td>
 </td></tr>, <tr>
 <td>M3B
 </td>
 <td>North York
 </td>
 <td>Do

In [7]:
# Find all table header names with <th> tag and create a list of column names
column_names = []
for ele in rows[0].find_all('th'):
    column_names.append(ele.get_text(strip=True))
column_names

['Postal Code', 'Borough', 'Neighborhood']

In [8]:
# Row by row, find all table cells with <td>, then create a list of all table cells
data = []
for row in rows[1:]:
    data_row = []
    for ele in row.find_all('td'):
        data_row.append(ele.get_text(strip=True))
    data.append(data_row)
data

[['M1A', 'Not assigned', ''],
 ['M2A', 'Not assigned', ''],
 ['M3A', 'North York', 'Parkwoods'],
 ['M4A', 'North York', 'Victoria Village'],
 ['M5A', 'Downtown Toronto', 'Regent Park, Harbourfront'],
 ['M6A', 'North York', 'Lawrence Manor, Lawrence Heights'],
 ['M7A', 'Downtown Toronto', "Queen's Park, Ontario Provincial Government"],
 ['M8A', 'Not assigned', ''],
 ['M9A', 'Etobicoke', 'Islington Avenue'],
 ['M1B', 'Scarborough', 'Malvern, Rouge'],
 ['M2B', 'Not assigned', ''],
 ['M3B', 'North York', 'Don Mills'],
 ['M4B', 'East York', 'Parkview Hill, Woodbine Gardens'],
 ['M5B', 'Downtown Toronto', 'Garden District, Ryerson'],
 ['M6B', 'North York', 'Glencairn'],
 ['M7B', 'Not assigned', ''],
 ['M8B', 'Not assigned', ''],
 ['M9B',
  'Etobicoke',
  'West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale'],
 ['M1C', 'Scarborough', 'Rouge Hill, Port Union, Highland Creek'],
 ['M2C', 'Not assigned', ''],
 ['M3C', 'North York', 'Don Mills'],
 ['M4C', 'East York', 'Woodbine 

In [9]:
# Transform the list of table cells into a dataframe
df = pd.DataFrame(data, columns=column_names)
df

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
26          M9C         Etobicoke   
27          M1E       Scarborough   
28          M2E      Not assigned   
29          M3E      Not assigned   
..          ...               ...   
150         M7W      Not assigned   
151         M8W         Etobicoke   
152         M9W         Etobicoke   
153         M1X       Scarborough   
154         M2X      Not assigned   
155         M3X      Not assigned   
156         M4X  Downtown Toronto   
157         M5X  Downtown Toronto   
158         M6X      Not assigned   
159         M7X      Not assigned   
160         M8X         Etobicoke   
161         M9X      Not assigned   
162         M1Y      Not assigned   
163         M2Y      Not assigned   
164         M3Y      Not assigned   
165         M4Y  Downtown Toronto   
166         M5Y      Not assigned   
167         M6Y      Not assigned   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
170         M9Y      Not assigned   
171         M1Z      Not assigned   
172         M2Z      Not assigned   
173         M3Z      Not assigned   
174         M4Z      Not assigned   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                          Neighborhood  
0                                                       
1                                                       
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
7                                                       
8                                     Islington Avenue  
9                                       Malvern, Rouge  
10                                                      
11                                           Don Mills  
12                     Parkview Hill, Woodbine Gardens  
13                            Garden District, Ryerson  
14                                           Glencairn  
15                                                      
16                                                      
17   West Deane Park, Princess Gardens, Martin Grov...  
18              Rouge Hill, Port Union, Highland Creek  
19                                                      
20                                           Don Mills  
21                                    Woodbine Heights  
22                                      St. James Town  
23                                  Humewood-Cedarvale  
24                                                      
25                                                      
26   Eringate, Bloordale Gardens, Old Burnhamthorpe...  
27                   Guildwood, Morningside, West Hill  
28                                                      

Clean the DataFrame so that:
* Rows with __Not assigned__ borough are dropped.
* Neighborhoods of the same postal code should be presented in one row, instead of separate rows of the same postal code.
* If a borough has a __Not assigned__ neighborhood, the neighborhood will have the same name as the borough.

In [10]:
# Remove all rows with 'Not assigned' borough
df = df[df['Borough'] != 'Not assigned'].reset_index(drop=True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
..          ...               ...   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
5                                     Islington Avenue  
6                                       Malvern, Rouge  
7                                            Don Mills  
8                      Parkview Hill, Woodbine Gardens  
9                             Garden District, Ryerson  
10                                           Glencairn  
11   West Deane Park, Princess Gardens, Martin Grov...  
12              Rouge Hill, Port Union, Highland Creek  
13                                           Don Mills  
14                                    Woodbine Heights  
15                                      St. James Town  
16                                  Humewood-Cedarvale  
17   Eringate, Bloordale Gardens, Old Burnhamthorpe...  
18                   Guildwood, Morningside, West Hill  
19                                         The Beaches  
20                                         Berczy Park  
21                                 Caledonia-Fairbanks  
22                                              Woburn  
23                                             Leaside  
24                                  Central Bay Street  
25                                            Christie  
26                                           Cedarbrae  
27                                   Hillcrest Village  
28     Bathurst Manor, Wilson Heights, Downsview North  

In [11]:
# Check that all values in the 'Postal Code' column are unique, i.e. neighborhoods of the same postal code are presented in one row only
df['Postal Code'].is_unique

True

In [12]:
# Check that no values in the 'Neighborhood' column are empty, i.e. no empty string ''
df['Neighborhood'].isin(['']).value_counts()

False    103
Name: Neighborhood, dtype: int64

In [13]:
df.head(12)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4         Queen's Park, Ontario Provincial Government  
5                                    Islington Avenue  
6                                      Malvern, Rouge  
7                                           Don Mills  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...

In [14]:
df.shape

(103, 3)

There are 103 rows in the dataframe.

# Part 2: Combining neighborhood data with geospatial data

Since geocoder.google function doesn't work, we are going to use the csv file to obtain geospatial data.

In [15]:
# Read data from CSV file
csv_path = 'https://cocl.us/Geospatial_data'
df_geospatial = pd.read_csv(csv_path)

In [16]:
df_geospatial.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [17]:
# Join dataframes df with df_geospatial with column 'Postal Code' as key
df = df.join(df_geospatial.set_index('Postal Code'), on='Postal Code', how='left')
df.head(12)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   

                                         Neighborhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5                                    Islington Avenue  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

# Part 3: Exploring and clustering

We will work with only boroughs that contain the word Toronto for further exploring and clustering analysis.

In [18]:
# Filter out only boroughs containing 'Toronto' to work with
df_toronto = df[df['Borough'].str.contains('Toronto', case=False, regex=False)]

# Reset index of df_toronto
df_toronto.reset_index(drop=True, inplace=True)
df_toronto

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                            Garden District, Ryerson  43.657162 -79.378937  
3                                      St. James Town  43.651494 -79.375418  
4                                         The Beaches  43.676357 -79.293031  
5                                         Berczy Park  43.644771 -79.373306  
6                                  Central Bay Street  43.657952 -79.387383  
7                                            Christie  43.669542 -79.422564  
8                            Richmond, Adelaide, King  43.650571 -79.384568  
9                        Dufferin, Dovercourt Village  43.669005 -79.442259  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21                           Forest Hill North & West  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                                 North Toronto West  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                     University of Toronto, Harbord  43.662696 -79.400049  
28                                 Runnymede, Swansea  43.651571 -79.484450  
29                        Moore Park, Summerhill East  43.689574 -79.383160  
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
31  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -

In [19]:
df_toronto.shape

(39, 5)

There are 39 postal codes whose boroughs contain the word Toronto.

As multiple neighborhoods can share the same postal code, and one borough can have multiple postal codes, we will use postal codes as the key for the following exploration.

Get the geographical coordinates of Toronto.

In [20]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="on_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


## Visualizing Toronto with the 39 postal codes

In [21]:
# Create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# Add markers to map
for lat, lng, label in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Postal Code']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## Exploring postal codes with Foursquare

Define Foursquare Credentials and Version

In [22]:
# The code was removed by Watson Studio for sharing.

Create a function to find the top 100 venues that are in a postal code within a radius of 500 metres.

In [23]:
def getNearbyVenues(postal_codes, latitudes, longitudes, radius=500, LIMIT=100):
    
    venues_list=[]
    for postal_codes, lat, lng in zip(postal_codes, latitudes, longitudes):
        print(postal_codes)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            postal_codes, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Postal Code', 
                  'Postal Code Latitude', 
                  'Postal Code Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Run the above function on each postal code and create a new dataframe called toronto_venues.

In [24]:
toronto_venues = getNearbyVenues(postal_codes=df_toronto['Postal Code'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )

M5A
M7A
M5B
M5C
M4E
M5E
M5G
M6G
M5H
M6H
M5J
M6J
M4K
M5K
M6K
M4L
M5L
M4M
M4N
M5N
M4P
M5P
M6P
M4R
M5R
M6R
M4S
M5S
M6S
M4T
M5T
M4V
M5V
M4W
M5W
M4X
M5X
M4Y
M7Y


Check the size of the resulting dataframe.

In [25]:
print(toronto_venues.shape)
toronto_venues.head()

(1606, 7)


Postal Code  Postal Code Latitude  Postal Code Longitude  \
0         M5A              43.65426             -79.360636   
1         M5A              43.65426             -79.360636   
2         M5A              43.65426             -79.360636   
3         M5A              43.65426             -79.360636   
4         M5A              43.65426             -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

Check how many venues were returned for each postal code.

In [26]:
toronto_venues.groupby('Postal Code').count()

Postal Code Latitude  Postal Code Longitude  Venue  \
Postal Code                                                       
M4E                             5                      5      5   
M4K                            42                     42     42   
M4L                            21                     21     21   
M4M                            40                     40     40   
M4N                             3                      3      3   
M4P                             8                      8      8   
M4R                            21                     21     21   
M4S                            33                     33     33   
M4T                             4                      4      4   
M4V                            16                     16     16   
M4W                             4                      4      4   
M4X                            49                     49     49   
M4Y                            74                     74     74   
M5A                            45                     45     45   
M5B                           100                    100    100   
M5C                            76                     76     76   
M5E                            56                     56     56   
M5G                            63                     63     63   
M5H                            93                     93     93   
M5J                           100                    100    100   
M5K                           100                    100    100   
M5L                           100                    100    100   
M5N                             3                      3      3   
M5P                             5                      5      5   
M5R                            23                     23     23   
M5S                            35                     35     35   
M5T                            56                     56     56   
M5V                            16                     16     16   
M5W                            94                     94     94   
M5X                           100                    100    100   
M6G                            16                     16     16   
M6H                            18                     18     18   
M6J                            41                     41     41   
M6K                            23                     23     23   
M6P                            22                     22     22   
M6R                            14                     14     14   
M6S                            36                     36     36   
M7A                            34                     34     34   
M7Y                            17                     17     17   

             Venue Latitude  Venue Longitude  Venue Category  
Postal Code                                                   
M4E                       5                5               5  
M4K                      42               42              42  
M4L                      21               21              21  
M4M                      40               40              40  
M4N                       3                3               3  
M4P                       8                8               8  
M4R                      21               21              21  
M4S                      33               33              33  
M4T                       4                4               4  
M4V                      16               16              16  
M4W                       4                4               4  
M4X                      49               49              49  
M4Y                      74               74              74  
M5A                      45               45              45  
M5B                     100              100             100  
M5C                      76               76              76  
M5E                      56               56              56  
M5G                      63               63              63  
M5H  

Find out how many unique categories can be curated from all the returned venues.

In [27]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 233 uniques categories.


### Analyzing each postal code

In [28]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add postal code column back to dataframe
toronto_onehot['Postal Code'] = toronto_venues['Postal Code'] 

# move postal code column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Postal Code  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0         M5A                  0        0                   0             0   
1         M5A                  0        0                   0             0   
2         M5A                  0        0                   0             0   
3         M5A                  0        0                   0             0   
4         M5A                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Toy / Game Store  Trail  Train Station  \
0             0  ...                 0      0              0   
1             0  ...                 0      0              0   
2             0  ...                 0      0              0   
3             0  ...                 0      0              0   
4             0  ...                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0         0            0              0            0  
1         0            0              0            0  
2         0            0              0            0  
3         0            0              0            0  
4         0            0              0            0  

[5 rows x 234 columns]

Examine the new dataframe size.

In [29]:
toronto_onehot.shape

(1606, 234)

Group rows by postal code and by taking the mean of the frequency of occurrence of each category.

In [30]:
toronto_grouped = toronto_onehot.groupby('Postal Code').mean().reset_index()
toronto_grouped

Postal Code  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0          M4E           0.000000   0.0000              0.0000        0.0000   
1          M4K           0.000000   0.0000              0.0000        0.0000   
2          M4L           0.000000   0.0000              0.0000        0.0000   
3          M4M           0.000000   0.0000              0.0000        0.0000   
4          M4N           0.000000   0.0000              0.0000        0.0000   
5          M4P           0.000000   0.0000              0.0000        0.0000   
6          M4R           0.000000   0.0000              0.0000        0.0000   
7          M4S           0.000000   0.0000              0.0000        0.0000   
8          M4T           0.000000   0.0000              0.0000        0.0000   
9          M4V           0.000000   0.0000              0.0000        0.0000   
10         M4W           0.000000   0.0000              0.0000        0.0000   
11         M4X           0.000000   0.0000              0.0000        0.0000   
12         M4Y           0.013514   0.0000              0.0000        0.0000   
13         M5A           0.000000   0.0000              0.0000        0.0000   
14         M5B           0.000000   0.0000              0.0000        0.0000   
15         M5C           0.000000   0.0000              0.0000        0.0000   
16         M5E           0.000000   0.0000              0.0000        0.0000   
17         M5G           0.000000   0.0000              0.0000        0.0000   
18         M5H           0.000000   0.0000              0.0000        0.0000   
19         M5J           0.000000   0.0000              0.0000        0.0000   
20         M5K           0.000000   0.0000              0.0000        0.0000   
21         M5L           0.000000   0.0000              0.0000        0.0000   
22         M5N           0.000000   0.0000              0.0000        0.0000   
23         M5P           0.000000   0.0000              0.0000        0.0000   
24         M5R           0.000000   0.0000              0.0000        0.0000   
25         M5S           0.000000   0.0000              0.0000        0.0000   
26         M5T           0.000000   0.0000              0.0000        0.0000   
27         M5V           0.000000   0.0625              0.0625        0.0625   
28         M5W           0.000000   0.0000              0.0000        0.0000   
29         M5X           0.000000   0.0000              0.0000        0.0000   
30         M6G           0.000000   0.0000              0.0000        0.0000   
31         M6H           0.000000   0.0000              0.0000        0.0000   
32         M6J           0.000000   0.0000              0.0000        0.0000   
33         M6K           0.000000   0.0000              0.0000        0.0000   
34         M6P           0.000000   0.0000              0.0000        0.0000   
35         M6R           0.000000   0.0000              0.0000        0.0000   
36         M6S           0.000000   0.0000              0.0000        0.0000   
37         M7A           0.000000   0.0000              0.0000        0.0000   
38         M7Y           0.000000   0.0000              0.0000        0.0000   

    Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0            0.000           0.0000             0.000             0.000000   
1            0.000           0.0000             0.000             0.023810   
2            0.000           0.0000             0.000             0.000000   
3            0.000           0.0000             0.000             0.050000   
4            0.000           0.0000             0.000             0.000000   
5            0.000           0.0000             0.000             0.000000   
6            0.000           0.0000             0.000             0.000000   
7            0.000           0.0000             0.000             0.000000   
8            0.000           0.0000             0.000             0.000000   
9            0.000    

Examine the new dataframe size.

In [31]:
toronto_grouped.shape

(39, 234)

Print each postal code along with the top 5 most common venues.

In [32]:
num_top_venues = 5

for hood in toronto_grouped['Postal Code']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Postal Code'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----M4E----
               venue  freq
0  Health Food Store   0.2
1   Asian Restaurant   0.2
2                Pub   0.2
3       Neighborhood   0.2
4              Trail   0.2


----M4K----
                    venue  freq
0        Greek Restaurant  0.21
1             Coffee Shop  0.07
2      Italian Restaurant  0.07
3  Furniture / Home Store  0.05
4          Ice Cream Shop  0.05


----M4L----
                  venue  freq
0           Pizza Place  0.10
1        Sandwich Place  0.10
2  Fast Food Restaurant  0.10
3      Sushi Restaurant  0.05
4                  Park  0.05


----M4M----
                 venue  freq
0                 Café  0.10
1          Coffee Shop  0.08
2              Brewery  0.05
3            Gastropub  0.05
4  American Restaurant  0.05


----M4N----
               venue  freq
0               Park  0.33
1           Bus Line  0.33
2        Swim School  0.33
3  Afghan Restaurant  0.00
4        Music Venue  0.00


----M4P----
                  venue  freq
0        Breakfast

Write a function to sort the venues in descending order.

In [33]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Create a new dataframe and display the top 10 venues for each postal code.

In [34]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Postal Code']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
postal_codes_venues_sorted = pd.DataFrame(columns=columns)
postal_codes_venues_sorted['Postal Code'] = toronto_grouped['Postal Code']

for ind in np.arange(toronto_grouped.shape[0]):
    postal_codes_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

postal_codes_venues_sorted.head()

Postal Code 1st Most Common Venue 2nd Most Common Venue  \
0         M4E      Asian Restaurant     Health Food Store   
1         M4K      Greek Restaurant    Italian Restaurant   
2         M4L        Sandwich Place  Fast Food Restaurant   
3         M4M                  Café           Coffee Shop   
4         M4N                  Park           Swim School   

  3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  \
0                   Pub                   Trail          Neighborhood   
1           Coffee Shop  Furniture / Home Store            Restaurant   
2           Pizza Place               Pet Store                   Pub   
3               Brewery               Gastropub                Bakery   
4              Bus Line             Yoga Studio        Discount Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0           Event Space  Ethiopian Restaurant     Electronics Store   
1        Ice Cream Shop           Yoga Studio       Bubble Tea Shop   
2          Liquor Store         Burrito Place     Fish & Chips Shop   
3   American Restaurant     Convenience Store    Seafood Restaurant   
4        Farmers Market    Falafel Restaurant           Event Space   

         9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant           Dessert Shop  
1                          Spa              Juice Bar  
2           Italian Restaurant             Steakhouse  
3               Sandwich Place            Cheese Shop  
4         Ethiopian Restaurant      Electronics Store

## Clustering postal codes

Run k-means to cluster the postal codes into 5 clusters.

In [35]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Postal Code', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 1, 1, 1, 2, 1, 1, 1, 1, 1], dtype=int32)

Create a new dataframe that includes the cluster as well as the top 10 venues for each postal code.

In [36]:
# add clustering labels
postal_codes_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(postal_codes_venues_sorted.set_index('Postal Code'), on='Postal Code')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               1           Coffee Shop   
1  43.662301 -79.389494               1           Coffee Shop   
2  43.657162 -79.378937               1        Clothing Store   
3  43.651494 -79.375418               1           Coffee Shop   
4  43.676357 -79.293031               0      Asian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0                  Park                Bakery                    Theater   
1      Sushi Restaurant   Arts & Crafts Store              Burrito Place   
2           Coffee Shop            Restaurant  Middle Eastern Restaurant   
3                  Café          Cocktail Bar                  Gastropub   
4     Health Food Store                   Pub                      Trail   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot                   Pub                  Café   
1                  Café         Smoothie Shop   Japanese Restaurant   
2       Bubble Tea Shop                  Café    Italian Restaurant   
3   American Restaurant                   Gym   Moroccan Restaurant   
4          Neighborhood           Event Space  Ethiopian Restaurant   

  8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
0     Electronics Store                   Beer Store                    Spa  
1              Beer Bar           Italian Restaurant     Mexican Restaurant  
2   Japanese Restaurant               Cosmetics Shop              Bookstore  
3        Cosmetics Shop             Department Store         Clothing Store  
4     Electronics Store  Eastern European Restaurant           Dessert Shop

### Visualizing the resulting clusters

In [37]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Postal Code'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examining clusters

#### Cluster 1

In [38]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[0] + list(range(6, toronto_merged.shape[1]))]]

Postal Code 1st Most Common Venue 2nd Most Common Venue  \
4         M4E      Asian Restaurant     Health Food Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4                   Pub                 Trail          Neighborhood   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4           Event Space  Ethiopian Restaurant     Electronics Store   

         9th Most Common Venue 10th Most Common Venue  
4  Eastern European Restaurant           Dessert Shop

#### Cluster 2

In [39]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[0] + list(range(6, toronto_merged.shape[1]))]]

Postal Code 1st Most Common Venue 2nd Most Common Venue  \
0          M5A           Coffee Shop                  Park   
1          M7A           Coffee Shop      Sushi Restaurant   
2          M5B        Clothing Store           Coffee Shop   
3          M5C           Coffee Shop                  Café   
5          M5E           Coffee Shop          Cocktail Bar   
6          M5G           Coffee Shop        Sandwich Place   
7          M6G         Grocery Store                  Café   
8          M5H           Coffee Shop                  Café   
9          M6H              Pharmacy                Bakery   
10         M5J           Coffee Shop              Aquarium   
11         M6J                   Bar            Restaurant   
12         M4K      Greek Restaurant    Italian Restaurant   
13         M5K           Coffee Shop                  Café   
14         M6K                  Café           Coffee Shop   
15         M4L        Sandwich Place  Fast Food Restaurant   
16         M5L           Coffee Shop            Restaurant   
17         M4M                  Café           Coffee Shop   
20         M4P  Gym / Fitness Center     Food & Drink Shop   
22         M6P    Mexican Restaurant                  Café   
23         M4R        Clothing Store           Coffee Shop   
24         M5R        Sandwich Place                  Café   
25         M6R        Breakfast Spot             Gift Shop   
26         M4S          Dessert Shop        Sandwich Place   
27         M5S                  Café                   Bar   
28         M6S      Sushi Restaurant                  Café   
29         M4T                  Park            Restaurant   
30         M5T                  Café           Coffee Shop   
31         M4V           Coffee Shop                   Pub   
32         M5V       Airport Service        Airport Lounge   
34         M5W           Coffee Shop    Seafood Restaurant   
35         M4X           Coffee Shop           Pizza Place   
36         M5X           Coffee Shop                  Café   
37         M4Y      Sushi Restaurant   Japanese Restaurant   
38         M7Y           Yoga Studio         Auto Workshop   

        3rd Most Common Venue      4th Most Common Venue  \
0                      Bakery                    Theater   
1         Arts & Crafts Store              Burrito Place   
2                  Restaurant  Middle Eastern Restaurant   
3                Cocktail Bar                  Gastropub   
5                        Café         Seafood Restaurant   
6          Italian Restaurant                       Café   
7                        Park                 Restaurant   
8                  Restaurant             Clothing Store   
9   Middle Eastern Restaurant                Music Venue   
10                       Café                      Hotel   
11      Vietnamese Restaurant                       Café   
12                Coffee Shop     Furniture / Home Store   
13                      Hotel                 Restaurant   
14             Breakfast Spot      Performing Arts Venue   
15                Pizza Place                  Pet Store   
16                       Café                      Hotel   
17                    Brewery                  Gastropub   
20           Department Store               Dance Studio   
22            Thai Restaurant                       Park   
23                Yoga Studio                      Diner   
24                Coffee Shop                  Pet Store   
25                  Bookstore                    Dog Run   
26                Pizza Place         Italian Restaurant   
27         Italian Restaurant        Japanese Restaurant   
28                Coffee Shop                Pizza Place   
29               Tennis Court                Summer Camp   
30      Vietnamese Restaurant         Mexican Restaurant   
31               Liquor Store                 Sports Bar   
32           Airport Terminal            Harbor / Marina   
34                       Café   

#### Cluster 3

In [40]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[0] + list(range(6, toronto_merged.shape[1]))]]

Postal Code 1st Most Common Venue 2nd Most Common Venue  \
18         M4N                  Park           Swim School   
21         M5P                  Park                 Trail   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
18              Bus Line           Yoga Studio        Discount Store   
21         Jewelry Store      Sushi Restaurant              Bus Line   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
18        Farmers Market    Falafel Restaurant           Event Space   
21           Yoga Studio        Discount Store    Falafel Restaurant   

   9th Most Common Venue 10th Most Common Venue  
18  Ethiopian Restaurant      Electronics Store  
21           Event Space   Ethiopian Restaurant

#### Cluster 4

In [41]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[0] + list(range(6, toronto_merged.shape[1]))]]

Postal Code 1st Most Common Venue 2nd Most Common Venue  \
33         M4W                  Park            Playground   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
33                 Trail           Yoga Studio          Dessert Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
33    Falafel Restaurant           Event Space  Ethiopian Restaurant   

   9th Most Common Venue       10th Most Common Venue  
33     Electronics Store  Eastern European Restaurant

#### Cluster 5

In [42]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[0] + list(range(6, toronto_merged.shape[1]))]]

Postal Code 1st Most Common Venue 2nd Most Common Venue  \
19         M5N                Garden          Home Service   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
19        Ice Cream Shop           Yoga Studio        Farmers Market   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
19    Falafel Restaurant           Event Space  Ethiopian Restaurant   

   9th Most Common Venue       10th Most Common Venue  
19     Electronics Store  Eastern European Restaurant